src/rpc/dial.rs

```diff
@@ -365,39 +379,132 @@ async fn get_auth_token(
+        }
+
+        let ip_addr = match resp.ip_addr() {
+            Some(std::net::IpAddr::V4(ip_v4)) => Some(ip_v4),
+            Some(std::net::IpAddr::V6(_)) | None => None,
+        };
+
+        if !(has_grpc || has_webrtc) || ip_addr.is_none() {
+            log::debug!("Unable to connect via mDNS");
+            return None;
+        }
+
+        let mut local_addr = ip_addr?.to_string();
+        local_addr.push(':');
+        local_addr.push_str(&resp.port()?.to_string());
+        log::debug!("Found address via mDNS: {local_addr}");
+
+        let auth = local_addr.parse::<Authority>().ok()?;
+        uri.authority = Some(auth);
+        uri.scheme = Some(Scheme::HTTP);
+
+        Some(uri)
+    }
+
+    async fn connect_inner(
+        self,
+        mdns_uri: Option<Parts>,
+        mut original_uri_parts: Parts,
+    ) -> Result<AddAuthorization<ViamChannel>> {
+        let is_insecure = self.config.insecure;
+
         let webrtc_options = self.config.webrtc_options;
         let disable_webrtc = match &webrtc_options {
             Some(options) => options.disable_webrtc,
             None => false,
         };
-        let mut uri_parts = self.config.uri.unwrap();
-        if self.config.insecure {
-            uri_parts.scheme = Some(Scheme::HTTP);
+
+        if is_insecure {
+            original_uri_parts.scheme = Some(Scheme::HTTP);
         }
 
-        let uri = Uri::from_parts(uri_parts)?;
-        let uri2 = uri.clone();
-        let domain = uri2.authority().to_owned().unwrap().to_string();
+        let original_uri = Uri::from_parts(original_uri_parts)?;
 
-        let uri = infer_remote_uri_from_authority(uri);
+        let domain = original_uri.authority().clone().unwrap().to_string();
+        let uri_for_auth = infer_remote_uri_from_authority(original_uri.clone());
 
-        let real_channel = match Channel::builder(uri.clone())
-            .connect()
-            .await
-            .with_context(|| format!("Connecting to {:?}", uri.clone()))
-        {
-            Ok(c) => c,
-            Err(e) => {
-                if self.config.allow_downgrade {
-                    let mut uri_parts = uri.clone().into_parts();
-                    uri_parts.scheme = Some(Scheme::HTTP);
-                    let uri = Uri::from_parts(uri_parts)?;
-                    Channel::builder(uri.clone()).connect().await?
-                } else {
-                    return Err(anyhow::anyhow!(e));
+        let mdns_uri = mdns_uri.and_then(|p| Uri::from_parts(p).ok());
+        let attempting_mdns = mdns_uri.is_some();
+
+        let allow_downgrade = self.config.allow_downgrade;
+        if attempting_mdns {
+            log::debug!("Attempting to connect via mDNS");
+        } else {
+            log::debug!("Attempting to connect");
+        }
+        let channel = match mdns_uri {
+            Some(uri) => Self::create_channel(allow_downgrade, &domain, uri, true)
+                .await
+                .ok(),
```

Comment:
I think we should keep the error here and report it on line 505, might be helpful for debugging mdns connection failure.

Reply:
ahh good catch! updated